scalableminds/webknossos
# Allow renaming datasets & dataset with duplicate names

### Further Notes:
- Quite some of the line changes are the result of moving the ObjectId class to the utils package so that all wk backend servers have access to this class.

### URL of deployed dev instance (used for testing):
- https://allowdatasetrenaming.webknossos.xyz

### Steps to test:
- Give two datasets the same name and check whether annotations and so on works
- Test whether the task system still works with duplicate dataset names
- check dataset upload
  - dataset upload
  - add remote
  - compose
- check moving datasets between folders (single & multiple at once)
- check whether the renaming works
- when there are duplicate names, the correct dataset should open upon view it (not another one with the same name)
- Test worker jobs (tested by @fm3)
- Test legacy routes
- And much much more :see_no_evil: 

### TODOs:
- [x] Add evolution and reversion
  - [x] testing needed
- [x] Test uploading: 
  - [x] Report upload fails
- [x] Adjust worker to newest job arguments as the dataset name can no longer be used to uniquely identify a dataset
- [x] rename `organization_name` in worker to `organization_id`. see #8038
- [x] Dataset Name settings field has an unwanted spinner (see upload view)
- [x] Check the job list
- [x] Properly implement legacy searching for datasets when old URI param is used 
- [x] Adjust legacy API routes to return dataset in old format
   - It is just an additional field. Thus, I would say it should be fine. 
- [x] datasets appear to be duplicated in the db 
  - Maybe these are created by jobs with an output dataset 
- [x] Fix dataset insert
- [x] Skeleton & VolumeTracings address a dataset via its name
  - Not really used only during task / annotation creation
  - Use heuristic upon upload and temporary patch the Tracing case classes to carry the datasetId during the creation process once the dataset is identified once.
  - Task creation works
  - [x] Needs testing
    - fix annotation upload 
  - [x] needs to support old nmls
- [x] Put datasetId into newly created nmls
- [x] In the backend `LinkedLayerIdentifier` still uses the datasetName as an identifier
  - used in wklibs, maybe just interpret the name as a path and work with this. in case it cannot be found the user needs to update wklibs. Add comment for this!
- ~~[ ] the dataset `C555_tps_demo` has quite some bucket loading errors. Unsure why some buckets do not work~~ The dataset seems to be broken. Could reproduce this on other branches
- [x] Notion-style URLs are missing (i.e. <name>-<id>, but only the id part is actually used)
- [x] Maybe remove `DatasetURIParser`
> I would also suggest to
> - [x] Execute the screenshot tests once for this branch. Take care to change the branch name in the CI config (or local command), so that the tests are actually ran on this branch.
- [x] Looks much better now :) Did you re-test that exceptions are caught as expected?
- [x] undo application.conf & snapshot changes -> in progress


### Issues:
- fixes #6613
- fixes #8038
- fixes #7560

------
(Please delete unneeded items, merge only when none are left open)
- [x] Needs to execute a mirgation
